api/C/include/csi_ll_pmp.h

```diff
@@ -0,0 +1,237 @@
+/*
+ * Low-Level PMP API
+ *
+ * This module provides functionality for PMP configuration.
+ *
+ * Copyright (C) 2017-2023 Alibaba Group Holding Limited
+ */
+
+#ifndef CSI_LL_PMP_H
+#define CSI_LL_PMP_H
+
+#include "csi_types.h"
+#include <stdint.h>
+
+/*
+ * Types of address Matching
+ */
+typedef enum {
+    ADDRESS_MATCHING_OFF   = 0,	/* Null region (disabled) */
+    ADDRESS_MATCHING_TOR   = 1, /* Top of range */
+    ADDRESS_MATCHING_NA4   = 2, /* Naturally aligned four-byte region */
+    ADDRESS_MATCHING_NAPOT = 3, /* Naturally aligned power-of-two region, ≥8 bytes */
+} csi_addr_matching_t;
+
+/*
+ * Attribution of PMP coniguration
+ */
+typedef struct {
+    bool flag_r; /* The flag of readable. 0: not readable; 1: readable */
+    bool flag_w; /* The flag of writable. 0: not writable; 1: writable */
+    bool flag_x; /* The flag of execable. 0: not execable; 1: execable */
+    csi_addr_matching_t mode; /* Address matching mode */
+    bool lock; /* The flag of lock,  0: not be locked; 1 : be locked */
+} csi_pmp_region_attr_t;
+
+/*
+ * Get PMP range number for the hard id.
+ *
+ * If the hard id is invalidate, such as out of support range, it should return
+ * an error
+ *
+ * @param hart_id: The hart id used to get the pmp range number
+ * @param range_number: A pointer to store the ouptut value of PMP range number
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_range_num(unsigned int hart_id, unsigned int *range_number);
```

Comment:
what do you mean by range number, does it mean the pmp entries supported? So this function is expected to do csr pmp register prope to get this entries number?

Reply:
Yes， it means the pmp entries supported. I think it could be implemented by doing csr pmp probe like opensbi or just a platform define like `#define PMP_COUNT			64`

Reply:
I think it would be better if this function returns the value instead of dereferencing a pointer.  Then the compiler can optimize it out, if the implementation in the BSP is simply a `#define`.

Reply:
> I think it would be better if this function returns the value instead of dereferencing a pointer. Then the compiler can optimize it out, if the implementation in the BSP is simply a `#define`.
Hi @WiseCrohn ，
My purpose to use error code is to avoid the case that invalid hart_id is input. So if I change this function to returns the value instead of dereferencing a pointer ,  is that suitable that it returns zero when invalid hart_id is input?

Reply:
I had written something along those lines replying to @WiseCrohn, I agree that it is simpler to just return but that implies that we have to decide and document what happens if the value for an invalid input (or other error) would be. I don't have a strong preference either way but, whenever possible we should try to be coherent in our approach.

Reply:
I think return an integer, and have -1 as an error code (e.g. invalid hart ID), and leave 0 to represent the case where there is no PMP (so there are no regions!).  This is analogous to what we have done elsewhere, e.g. csi_get_irq_priority etc.

Reply:
fixed. I've changed this API.  It wil return the value of PMP entry of specified hart id, and will return -1 if the input is invalid